File: src/main/java/xyz/baddeveloper/lwsl/client/SocketClient.java
package xyz.baddeveloper.lwsl.client;

import dev.throwouterror.eventbus.SimpleEventManager;
import org.json.JSONObject;
import xyz.baddeveloper.lwsl.Packet;
import xyz.baddeveloper.lwsl.client.events.*;
import xyz.baddeveloper.lwsl.exceptions.ConnectException;

import javax.net.ssl.SSLContext;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.Socket;
import java.util.concurrent.Executors;

public class SocketClient {

    private String address;
    private int port;
    private int timeout;
    private boolean keepalive;

    /*
        private List<OnConnectEvent> connectEvents = new ArrayList<>();
        private List<OnDisconnectEvent> disconnectEvents = new ArrayList<>();
        private List<OnPacketReceivedEvent> packetReceivedEvents = new ArrayList<>();
        private List<OnPacketSentEvent> packetSentEvents = new ArrayList<>();
    */
    private final SimpleEventManager eventManager = new SimpleEventManager();
    private boolean connected = false;
    private Socket socket;
    private final SSLContext sslContext;

    private DataInputStream dis;
    private DataOutputStream dos;

    public SocketClient(String address, int port) {
        this(address, port, null);
    }

    public SocketClient(String address, int port, SSLContext sslContext) {
        this.address = address;
        this.port = port;
        this.sslContext = sslContext;
    }

    public void connect() throws ConnectException {
        try {
            if (sslContext != null) {
                socket = sslContext.getSocketFactory().createSocket(address, port);
            } else {
                socket = new Socket(address, port);
            }

            socket.setKeepAlive(keepalive);
            socket.setSoTimeout(timeout);

            dis = new DataInputStream(socket.getInputStream());
            dos = new DataOutputStream(socket.getOutputStream());

        } catch (Exception e) {
            throw new ConnectException(e);
        }
        connected = socket != null;
        if (!connected) return;
        eventManager.fireEvent(new ClientConnectEvent(this));
        listen();
    }

    public void shutdown() {
        try {
            if (socket != null) socket.close();
            connected = false;
        } catch (IOException ignored) {
        }
    }

    private void listen() {
        Executors.newSingleThreadExecutor().execute(() -> {
            while (!socket.isClosed()) {
                try {
                    Packet in = new Packet(new JSONObject(dis.readUTF()));
                    eventManager.fireEvent(new ClientPacketReceivedEvent(this, in));
                } catch (Exception e) {
                    try {
                        socket.close();
                    } catch (IOException ignored) {
                    }
                    eventManager.fireEvent(new ClientDisconnectEvent(this));
                    break;
                }
            }
        });
        eventManager.fireEvent(new ClientReadyEvent(this));
    }

    public void sendPacket(Packet packet) {
        try {
            packet.setName(packet.getClass().getSimpleName().replace("Packet", ""));
            dos.writeUTF(packet.getObject().toString());
            dos.flush();
            eventManager.fireEvent(new ClientPacketSentEvent(this, packet));
        } catch (IOException e) {
            e.printStackTrace();
            //TODO: Create exception
        }
    }

    public SocketClient setAddress(String address) {
        this.address = address;
        return this;
    }

    public SocketClient setPort(int port) {
        this.port = port;
        return this;
    }

    public SocketClient setTimeout(int timeout) {
        this.timeout = timeout;
        return this;
    }

    public SocketClient setKeepAlive(boolean keepalive) {
        this.keepalive = keepalive;
        return this;
    }

    public SocketClient addEventListener(Object eventHandlerClass) {
        eventManager.addEventListener(eventHandlerClass);
        return this;
    }

    public boolean isKeepAlive() {
        return keepalive;
    }

    public boolean isConnected() {
        return connected;
    }

    public Socket getSocket() {
        return socket;
    }

    public String getAddress() {
        return address;
    }

    public int getPort() {
        return port;
    }

    public int getTimeout() {
        return timeout;
    }
}

File: src/main/java/xyz/baddeveloper/lwsl/client/SocketClient.kt
package xyz.baddeveloper.lwsl.client

import dev.throwouterror.eventbus.TypedEventBus
import io.netty.handler.logging.LogLevel
import org.json.JSONObject
import reactor.core.publisher.Mono
import reactor.netty.Connection
import reactor.netty.tcp.SslProvider
import reactor.netty.tcp.TcpClient
import xyz.baddeveloper.lwsl.Events
import xyz.baddeveloper.lwsl.client.events.ClientEvent
import xyz.baddeveloper.lwsl.client.events.ClientPacketEvent
import xyz.baddeveloper.lwsl.exceptions.ConnectException
import xyz.baddeveloper.lwsl.packet.Packet
import java.io.IOException
import javax.net.ssl.SSLContext

class SocketClient @JvmOverloads constructor(var address: String, var port: Int, private val sslProvider: SslProvider? = null) {
    var timeout = 0
        private set
    var isKeepAlive = false
        private set

    private val eventManager = TypedEventBus<ClientEvent>()
    var isConnected = false
        private set
    var client: TcpClient = TcpClient.create().host(address).port(port)
        private set
    lateinit var connection: Connection

    @Throws(ConnectException::class)
    fun connect(): SocketClient {
        listen()
        return this
    }

    fun shutdown() {
        try {
            connection.disposeNow()
            isConnected = false
        } catch (ignored: IOException) {
        }
    }

    private fun listen() {
        client.doOnConnected {
            eventManager.fireEvent(ClientEvent(Events.READY.toString(), this))
            try {
                val packetIn = Packet(JSONObject(it.inbound().receiveObject().subscribe()))
                eventManager.fireEvent(ClientPacketEvent(Events.PACKET_RECEIVED.toString(), this, packetIn))
            } catch (e: Exception) {
                try {
                    connection.disposeNow()
                } catch (ignored: IOException) {
                }
                eventManager.fireEvent(ClientEvent(Events.DISCONNECT.toString(), this))
                return@doOnConnected
            }
        }
        connection = client.connectNow()
    }

    fun sendPacket(packet: Packet) {
        try {
            packet.name = packet.javaClass.simpleName.replace("Packet", "")
            connection.outbound().sendString(Mono.just(packet.jsonObject.toString()))
            eventManager.fireEvent(ClientPacketEvent(Events.PACKET_SENT.toString(), this, packet))
        } catch (e: IOException) {
            e.printStackTrace()
            // TODO: Create exception
        }
    }

    fun enableDebugLogging(): SocketClient {
        client.wiretap("LWSL", LogLevel.INFO)
        return this
    }

    fun setAddress(address: String): SocketClient {
        this.address = address
        return this
    }

    fun setPort(port: Int): SocketClient {
        this.port = port
        return this
    }

    fun setTimeout(timeout: Int): SocketClient {
        this.timeout = timeout
        return this
    }

    fun setKeepAlive(keepalive: Boolean): SocketClient {
        isKeepAlive = keepalive
        return this
    }

    fun createEventBus(): TypedEventBus<ClientEvent> {
        return eventManager
    }

}